milli/src/search/query_tree.rs

```diff
@@ -476,13 +484,15 @@ impl PrimitiveQueryPart {
 
 /// Create primitive query from tokenized query string,
 /// the primitive query is an intermediate state to build the query tree.
-fn create_primitive_query(query: TokenStream, stop_words: Option<Set<&[u8]>>) -> PrimitiveQuery {
+fn create_primitive_query(query: TokenStream, stop_words: Option<Set<&[u8]>>, words_limit: Option<usize>) -> PrimitiveQuery {
     let mut primitive_query = Vec::new();
     let mut phrase = Vec::new();
     let mut quoted = false;
 
+    let parts_limit = words_limit.unwrap_or(usize::MAX);
+
     let mut peekable = query.peekable();
-    while let Some(token) = peekable.next() {
+    while let (Some(token), true) = (peekable.next(), primitive_query.len() < parts_limit) {
```

Comment:
```suggestion
    while let Some(token) = peekable.next() {
        // early return if word limit is exceded
        if primitive_query.len() >= parts_limit { return primitive_query }
```
I find it clearer and allow you to rollback Line 526 🙂